### PR TITLE
C#: Fix editor integration breaking and causing error spam when reloading assemblies fails

### PR DESCRIPTION
**Pull request by RedworkDE**
- Fixes the error spam of https://github.com/godotengine/godot/issues/70026

The core issue is, that reloading user assemblies also reloads GodotTools. Then when the reloading fails all the callbacks registered from GodotTools are invalid and cause the errors and menu items and buttons no longer work.

To fix the issue, this implements a TODO about not reloading scripts (and Callables) from non-collectible assemblies and loads GodotTools as non-collectible (it can never change without an editor update anyways).

This is enough to fix the signal error spam, but because the timer to reload assembly now no longer breaks when assembly loading fails, the assembly reloading will now be reattempted forever spamming the `.NET: Failed to load project assembly.` error instead. Thus this PR also limits this to at most 5 attempts to reload the same project assembly. (If we give up on loading the assembly and the user changes things and rebuild, which now actually works from the editor, we will attempt to load it 5 times as well or maybe load it successfully)

Also upgrades the `.NET: Failed to load project assembly.` error to an editor message as otherwise there is very little indication for the user that anything has failed.

Not marking this as fixing the issue, as I haven't even started to look into why unloading even fails in the first place as that issue doesn't even have an MRP. Instead I used that the editor won't load preview versions of .net atm to produce a project that cannot be loaded: [Test_70026.zip](https://github.com/godotengine/godot/files/11122085/Test_70026.zip) (This project required a .NET 8 preview installed to build, switching the target framework between net6.0 and net8.0 can be used to switch between a loadable and non-loadable project)

EDIT: ~This does actually fix the MRP from https://github.com/godotengine/godot/issues/72435, because while the first attempt at reloading fails, the editor integration no longer breaks at it does a second attempt, which then works (IMO the unloading should happen later after the script where unloaded to properly fix this, but that is for another PR)~

### CODE REVIEW

**Comment by magian1127** [comment]
Rebuilding the project in VS results in the `Another resource is loaded from path 'res://xxx.cs' (possible cyclic resource inclusion).` error, which causes `[Export]` to not update correctly.

**Comment by RedworkDE** [comment]
Pretty sure that the resource loaded error predates this PR and this just gives it more opportunities to pop up. I'll take a look at it later fix it here or in another PR, depending on how it turns out.

**Comment by RedworkDE** [comment]
I looked into things a bit more and it turns out unloading actually doesn't succeed at the second attempt, we just lost the reference to the old ALC and then load the project again without the old one being cleared out properly, which may be worth doing intentionally, but that should be a dedicated PR and discussion if we want to do that. Instead I make sure to preserve the weak reference to the ALC until it is actually unloaded.

Also because we now give it 5 attempts at reloading the assembly I reduced the unload timer to 1 second, so that we wait 5 * 1s instead of 5s once like before.

This "fixes" the apparent regressions from this PR by not reloading the assemblies when unloading fails.

RFC: I was going to open an issue about the causes of unloading failing and then link it from the error message about it so that users can get at starting point to look into this, because while the issue seems to have various ways which can trigger this issue, both MRPs actually fail to unload because they use third-party libraries that are not unloadable and the problems actually have nothing to do with Godot.
Is this something that should should be done / is there any preference for handling such a situation? (Using an issue over a docs page, to allow for easier discussions and to collect problematic libraries and actual problems with godot (if any))

**Comment by YuriSizov** [comment]
> I was going to open an issue about the causes of unloading failing and then link it from the error message about it so that users can get at starting point to look into this, because while the issue seems to have various ways which can trigger this issue, both MRPs actually fail to unload because they use third-party libraries that are not unloadable and the problems actually have nothing to do with Godot.
> Is this something that should should be done / is there any preference for handling such a situation? (Using an issue over a docs page, to allow for easier discussions and to collect problematic libraries and actual problems with godot (if any))

I think it's fine to make a tracker for that and to leave a reference to the issue somewhere in the error log. I would probably make sure we only print this warning in editor builds, though. So it doesn't leak in logs created by projects and direct users of games made with Godot this way by mistake.

**Review by YuriSizov — APPROVED** [review]
I haven't tested it, but your explanation makes a lot of sense to me. I'm okay merging it as a stepping stone for now, but perhaps @raulsntos or @neikeq could give a look to the code as well before we do that?

**Comment by dannymate** [comment]
The issue referenced #70026 now has an MRP for 4.1 beta 1. Although it seems that it's harder to reproduce in 4.1 beta 2 & 4.0 but hopefully that's enough. If not I'll see if I can figure out another one.

File: modules/mono/mono_gd/gd_mono.cpp
Comment (raulsntos) [inline modules/mono/mono_gd/gd_mono.cpp:515]: Should be able to use `path::get_csharp_project_name` here.

```suggestion
		String assembly_name = path::get_csharp_project_name();
```

File: modules/mono/mono_gd/gd_mono.cpp
Comment (raulsntos) [inline modules/mono/mono_gd/gd_mono.cpp:505]: Why are we attempting to reload 5 times? If it fails once why would it work the next time?

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs:30]: Here you return `true` when the target is null, but if the target is a `Delegate` the method `DelegateUtils.IsDelegateCollectible` returns `false` if the delegate is null. Obviously if the delegate is null it won't every reach the `DelegateUtils.IsDelegateCollectible` method invocation since it will return here. I was wondering why the 2 methods treat null differently.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/DelegateUtils.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/DelegateUtils.cs:578]: Does the `delegate.Target` type being collectible mean the delegate is also collectible even if the `delegate.Method` isn't?

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs:36]: Can the target be collectible if the type is not in a collectible assembly?

File: modules/mono/csharp_script.cpp
Comment (raulsntos) [inline modules/mono/csharp_script.cpp:873]: I don't understand what this is saying, is there a typo?

File: modules/mono/mono_gd/gd_mono.cpp
Comment (raulsntos) [inline modules/mono/mono_gd/gd_mono.cpp:529]: If we're going to retry loading multiple times, I feel like showing errors while we're still trying may be confusing to users. If reloading is able to succeed after multiple tries, then maybe we should wait until all the tries fail to show an error.

Currently with all these `ERR_PRINT_ED`, in the worst case scenario you get 6 error messages and that may be excessive.

File: modules/mono/mono_gd/gd_mono.cpp
Comment (raulsntos) [inline modules/mono/mono_gd/gd_mono.cpp:546]: If it was successful, should we really show an error message?

**Review by raulsntos — CHANGES_REQUESTED** [review]
I don't fully understand how collectible assemblies work so TIWAGOS, but the code makes sense to me for the most part. I made a few comments about changes that I don't understand.

Also, I'm a little concerned about retrying loading assemblies 5 times and the error editor messages.

File: modules/mono/mono_gd/gd_mono.cpp
Comment (RedworkDE) [reply modules/mono/mono_gd/gd_mono.cpp:505]: Tbh it has been too long I don't remember exactly why, but IIRC the design of assembly reloading means that currently, if the editor integration wouldn't break it would retry forever and coupled with the first version of this PR loosing the reference to the old assembly and appearing to work on the second attempt, I just left the retry behavior as-is and put a cap on it.

It could theoretically work after a few attempts if the current unloading was still doing something that just took more time and wasn't blocked completely, but in most cases unloading will succeed in maybe 10 GCs or it will never work.

File: modules/mono/mono_gd/gd_mono.cpp
Comment (RedworkDE) [reply modules/mono/mono_gd/gd_mono.cpp:529]: The 5 errors caused by this will get folded into a single popup so you really only get 2 editor popups.

File: modules/mono/mono_gd/gd_mono.cpp
Comment (RedworkDE) [reply modules/mono/mono_gd/gd_mono.cpp:546]: This only appears if the unloading has failed at least once before and thus an error popup about unloading failing has already been shown to the user, this message is only to let them know that the previous error no longer matters. It could probably be an success or info message instead, except that the first doesn't exist and the second doesn't have a convenient macro for spawning one.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs
Comment (RedworkDE) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs:36]: Yes, for example if the delegate is of type `Action`, then the type will not be reloadable as it is in the corelib, but the target and method can be in a reloadable assembly.

File: modules/mono/csharp_script.cpp
Comment (RedworkDE) [reply modules/mono/csharp_script.cpp:873]: *were saved in the // Serialize managed callables block

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs
Comment (RedworkDE) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs:30]: TBH this should be unreachable anyways, as it means that the passed handle was already released / collected already, also these checks are really only there to ignore things from the GodotTools assembly and the check generally should try to reload things to be on the safe side.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/DelegateUtils.cs
Comment (RedworkDE) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/DelegateUtils.cs:578]: If `A.foo()` is in a non-collectible assembly and `B` in a collectible assembly inherits from `A` and doesn't override `foo`, a delegate to `B.foo()` should have a collectible target but not a collectible method. And then it should be possible to release the delegate itself (which may or may not be of a collectible type itself) and make the instance of `B` available to be collected and thus possibly unblock the unloading of `B`.

Anyways these checks generally default to releasing the target to be safe and only try to prevent the unloading from things in GodotTools

File: modules/mono/mono_gd/gd_mono.cpp
Comment (raulsntos) [reply modules/mono/mono_gd/gd_mono.cpp:529]: You are right, and even though they all show individually in the output dock, it's probably fine.

File: modules/mono/mono_gd/gd_mono.cpp
Comment (raulsntos) [reply modules/mono/mono_gd/gd_mono.cpp:546]: Ah I see, it makes sense then.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs
Comment (raulsntos) [reply modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/GCHandleBridge.cs:30]: > TBH this should be unreachable anyways, as it means that the passed handle was already released / collected already

That's what I was thinking, but I figure it's safe to return `true` in this case. What concerns me is why this similar case returns `false` in:

https://github.com/RedworkDE/godot/blob/f4e115668625fcbd543dca293ecf48c402904b64/modules/mono/glue/GodotSharp/GodotSharp/Core/DelegateUtils.cs#L550-L553

Shouldn't it also return `true`? Although I realize I'm being nit-picky.

File: modules/mono/mono_gd/gd_mono.cpp
Comment (raulsntos) [reply modules/mono/mono_gd/gd_mono.cpp:505]: > It could theoretically work after a few attempts if the current unloading was still doing something that just took more time and wasn't blocked completely

That's what I had in mind, but I'm not sure how likely this scenario is.

>  I just left the retry behavior as-is and put a cap on it.

That makes sense, I was only suggesting to go even further and make the cap 1 instead of 5. But I guess it's safer to make multiple attempts just in case.

**Comment by RedworkDE** [comment]
> I think it's fine to make a tracker for that and to leave a reference to the issue somewhere in the error log. I would probably make sure we only print this warning in editor builds, though. So it doesn't leak in logs created by projects and direct users of games made with Godot this way by mistake.

-> https://github.com/godotengine/godot/issues/78513 The whole assembly unloading is editor only anyways and not compiled into the templates (`GD_MONO_HOT_RELOAD` is basically an alias for `TOOLS_ENABLED && MODULE_MONO_ENABLED`)

File: modules/mono/mono_gd/gd_mono.cpp
Comment (RedworkDE) [reply modules/mono/mono_gd/gd_mono.cpp]: I made this configurable as a project setting and reduced the default to 3 reload attempts.

File: modules/mono/mono_gd/gd_mono.cpp
Comment (raulsntos) [inline modules/mono/mono_gd/gd_mono.cpp:492]: This seems to be what we do in other usages of `GLOBAL_GET`:

```suggestion
	if (++project_load_failure_count >= (int)GLOBAL_GET("dotnet/project/assembly_reload_attempts")) {
```

But take it with a grain of salt, I'm not well versed in C++.

File: modules/mono/csharp_script.cpp
Comment (raulsntos) [inline modules/mono/csharp_script.cpp:922]:
```suggestion
	// Release the delegates that were serialized earlier.
```

File: modules/mono/csharp_script.cpp
Comment (raulsntos) [inline modules/mono/csharp_script.cpp:794]:
```suggestion
				// Cast to CSharpScript to avoid being erased by accident.
```

File: modules/mono/mono_gd/gd_mono.cpp
Comment (RedworkDE) [reply modules/mono/mono_gd/gd_mono.cpp:492]: `static_cast` is the "correct" c++ way of calling conversion functions like this, `(type)` are C style casts, which "should" generally be avoided as it can accidentally cast away `const` and it is not always clear how the cast is supposed to work.

Tho even when using c style casts, `int64_t` is the correct type or you risk accidental overflows (Variant stores 64 bit ints)

Comment (akien-mga) [reply modules/mono/mono_gd/gd_mono.cpp:492]: In this case, we value consistency over absolute correctness. There's no risk of casting away `const` here so a C style cast is fine and the most common usage in the Godot codebase.

File: modules/mono/csharp_script.cpp
Comment (raulsntos) [inline modules/mono/csharp_script.cpp:123]: Should this be an Editor setting instead?

```c++
EDITOR_DEF("dotnet/project/assembly_reload_attempts", 3);
EditorSettings::get_singleton()->add_property_hint(PropertyInfo(Variant::INT, "dotnet/project/assembly_reload_attempts", PROPERTY_HINT_RANGE, "1,16,1,or_greater"));
```

**Review by raulsntos — APPROVED** [review]
LGTM. May want to also get a review from @neikeq since this touches parts that I don't fully understand.

File: modules/mono/csharp_script.cpp
Comment (RedworkDE) [reply modules/mono/csharp_script.cpp:123]: Could be either, but I was thinking that if you have a reason to change this value, it is likely because something in the project takes a long time to unload and thus for that project only it should be increased by a bunch.

Comment (raulsntos) [reply modules/mono/csharp_script.cpp:123]: I guess that makes sense. I wish we could override editor settings per project (https://github.com/godotengine/godot-proposals/issues/1480).

**Comment by YuriSizov** [comment]
Between the two of you, I think we're safe merging it as is, at least for now. Thanks!

**Pull request merged**
